### PR TITLE
feat: instantiate solana

### PR DESCRIPTION
**Pull request by ramizhasan111**
# Pull Request

Closes: PRO-1329

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

This PR includes the following tasks:

- define sol types and impl them
- instantiate pallets for sol using correct types
- add sol storage items in env pallet
- migrations
- address derivation for sol
- impl Chain and ChainCrypto traits for sol
- other sol related required chores

#### Non-Breaking changes

If this PR includes non-breaking changes, select the `non-breaking` label. On merge, CI will automatically cherry-pick the commit to a PR against the release branch.

### CODE REVIEW

File: foreign-chains/solana/sol-prim/src/pda.rs
Comment (albert-llimos) [inline foreign-chains/solana/sol-prim/src/pda.rs:29]: As we just discussed, this address derivation seems to be an "optimized" version from the `PubKey::try_find_program_address()` used normally in Solana:
https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.try_find_program_address
We really need to make sure that we are not missing anything if we don't use the original code as is. I'd definitely prefer to implement the try_find_program_address as is instead of trying to reinvent the wheel. The upside is to generate deposit addresses slightly more efficiently. The downside is that we have a bug in address derivation that can end up in a catastrophic event.
If we look deep into it and this is a correct simplification of that code and we are sure about that, then it's alright I guess.

File: state-chain/runtime/src/migrations/arbitrum_integration.rs
Comment (ramizhasan111) [reply state-chain/runtime/src/migrations/arbitrum_integration.rs:1]: yeah I was going to do that

File: foreign-chains/solana/sol-prim/src/pda.rs
Comment (dandanlen) [reply foreign-chains/solana/sol-prim/src/pda.rs:29]: It looks like a sensible (and pretty simple) optimisation: it just avoids re-hashing all the seeds for every bump. If we're uncertain we can just copy the Solana tests and check it against those.

Comment (albert-llimos) [reply foreign-chains/solana/sol-prim/src/pda.rs:29]: okay, then we go with that :+1:

File: state-chain/runtime/src/lib.rs
Comment (dandanlen) [inline state-chain/runtime/src/lib.rs:1430]: Probably needs to be set somewhere? Would be good to have a list of pallet configuration items that need to be set for new chains...

**Review by dandanlen — COMMENTED** [review]
Nice, looks mostly fine to me.

File: state-chain/chains/src/sol/tracked_data.rs
Comment (albert-llimos) [reply state-chain/chains/src/sol/tracked_data.rs:18]: Pretty sure it's leftovers from the initial implementation. Should be removed.
And if I'm not mistaken this file is wrong, as we will not be tracking ingress/egress in Chain Tracking. I already have an implementation of this, so in this PR I'd remove the logic and leave the functions as todo.

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (albert-llimos) [reply state-chain/runtime/src/chainflip/address_derivation/sol.rs:41]: I think  the Solana convention is little endian, and that's also what Borsh encoder uses. So I'd use that.

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (albert-llimos) [reply state-chain/runtime/src/chainflip/address_derivation/sol.rs:43]: I think it's more "legacy" code. Ramiz can confirm.

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (albert-llimos) [inline state-chain/runtime/src/chainflip/address_derivation/sol.rs:40]: We have not decided yet on the best/default way to generate deposit addresses but this seed (`VAULT_PDA_ASSET_SOL`) is unnecessary, we can just derive it from the `channel_id`. We will maybe tweak this in the future but for now I'd remove it, as afaik there is no reason to do it and it will cost us more fees for no reason.

File: state-chain/chains/src/sol/tracked_data.rs
Comment (ramizhasan111) [reply state-chain/chains/src/sol/tracked_data.rs:18]: meant to remove it, forgot. removed now.

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (ramizhasan111) [reply state-chain/runtime/src/chainflip/address_derivation/sol.rs:43]: removed

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (ramizhasan111) [reply state-chain/runtime/src/chainflip/address_derivation/sol.rs:40]: done!

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (ramizhasan111) [reply state-chain/runtime/src/chainflip/address_derivation/sol.rs:41]: changed to little endian. It shouldnt matter whether we use le or be as long as we stick to our choice though right?

File: state-chain/cfe-events/src/tests.rs
Comment (ramizhasan111) [reply state-chain/cfe-events/src/tests.rs]: yeah, i thought about doing so but didnt like to break the symmetry :)
changed now.
regarding the sol checks, I had created an issue for this earlier (pro-1331) to do this in a separate PR.

File: foreign-chains/solana/sol-prim/src/pda.rs
Comment (ramizhasan111) [reply foreign-chains/solana/sol-prim/src/pda.rs:29]: double checked the impl, looks good.

File: state-chain/runtime/src/lib.rs
Comment (ramizhasan111) [reply state-chain/runtime/src/lib.rs:1435]: This has to be set via governance extrinsic ```update_pallet_config``` in the ingress egress pallet.

File: state-chain/chains/src/sol.rs
Comment (ramizhasan111) [reply state-chain/chains/src/sol.rs:41]: It is not immediately obvious which ones we will need. I will add a todo comment in front each of the empty types so that we can revisit them once all of the sol code is plugged in.

File: state-chain/runtime/src/chainflip/address_derivation/sol.rs
Comment (albert-llimos) [reply state-chain/runtime/src/chainflip/address_derivation/sol.rs]: technically yes. We would just be adding work to ensure we pass it correctly to the Vault program, which also does the same derivation. Let's not add extra trouble to ourselves :grin:

File: state-chain/runtime/src/lib.rs
Comment (dandanlen) [reply state-chain/runtime/src/lib.rs:1436]: Can we maintain a list of these tasks please, so that we don't forget?

Comment (ramizhasan111) [reply state-chain/runtime/src/lib.rs:1436]: sure, I will try to put together a list.

**Review by dandanlen — APPROVED** [review]

**Comment by dandanlen** [comment]
Please don't merge just yet - I would like us to consider how to manage this wrt. the 1.4 release and whether or not we branch 1.4 before merging this to main. 

Otherwise this looks good, thanks!

File: state-chain/runtime/src/lib.rs
Comment (ramizhasan111) [reply state-chain/runtime/src/lib.rs:1436]: created a notion doc for chain initialization realted tasks:

https://www.notion.so/chainflip/New-chain-initialization-todo-list-6c03c525fc9a455297e67c6d37c8bbf1

please add if anything else we need to do for new chains.

**Review by kylezs — CHANGES_REQUESTED** [review]
Upgrade test is failing migrations: https://github.com/chainflip-io/chainflip-backend/actions/runs/9212910929/job/25346016507?pr=4763

Also the versions should be bumped a minor version to 1.5.0

Lmk if you need a hand testing the migrations.

File: state-chain/runtime/src/migrations/solana_integration.rs
Comment (kylezs) [inline state-chain/runtime/src/migrations/solana_integration.rs:94]: do we not have the contract on testnets?

Comment (ramizhasan111) [reply state-chain/runtime/src/migrations/solana_integration.rs:94]: i dont think so. @albert-llimos ?

Comment (albert-llimos) [reply state-chain/runtime/src/migrations/solana_integration.rs:94]: No. Contracts are not finalized yet.

**Review by kylezs — APPROVED** [review]

**Pull request merged**
